File: pipeline-template/tests/testthat/test_pipeline.R
# tests/testthat/test_pipeline.R
#
# Unit tests for R functions in this pipeline.
# Run with: Rscript -e "testthat::test_dir('tests/testthat')"
#
# These tests cover pure logic only — no GCP connections.
# Functions that call bigrquery or googleCloudStorageR should be tested
# with mocked responses or in a separate integration test suite.

library(testthat)


# ---------------------------------------------------------------------------
# Example: testing a data cleaning function
# ---------------------------------------------------------------------------

clean_column_names <- function(df) {
  names(df) <- tolower(gsub("[^a-zA-Z0-9]", "_", names(df)))
  df
}

test_that("clean_column_names lowercases names", {
  df <- data.frame(MyColumn = 1, AnotherCol = 2)
  result <- clean_column_names(df)
  expect_equal(names(result), c("mycolumn", "anothercol"))
})

test_that("clean_column_names replaces spaces with underscores", {
  df <- data.frame(check.names = FALSE, "My Column" = 1)
  result <- clean_column_names(df)
  expect_equal(names(result), "my_column")
})

test_that("clean_column_names does not alter row count", {
  df <- data.frame(A = 1:5, B = 6:10)
  result <- clean_column_names(df)
  expect_equal(nrow(result), 5)
})

File: pipeline-template/tests/test_pipeline.py
# tests/test_pipeline.py
#
# Unit tests for Python functions in this pipeline.
# Run with: pytest tests/
#
# These tests cover pure logic only — no GCP connections.
# Anything that touches BigQuery or GCS should be tested with mocks
# or in an integration test suite run separately.

import pytest
import pandas as pd


# ---------------------------------------------------------------------------
# Example: testing a transformation function
# ---------------------------------------------------------------------------

def add_processed_timestamp(df: pd.DataFrame) -> pd.DataFrame:
    """Example transform function from src/transform.py."""
    df = df.copy()
    df["processed_at"] = pd.Timestamp.utcnow()
    return df


def test_add_processed_timestamp_adds_column():
    df = pd.DataFrame({"id": [1, 2], "value": [10, 20]})
    result = add_processed_timestamp(df)
    assert "processed_at" in result.columns


def test_add_processed_timestamp_preserves_rows():
    df = pd.DataFrame({"id": [1, 2, 3]})
    result = add_processed_timestamp(df)
    assert len(result) == 3


def test_add_processed_timestamp_does_not_mutate_input():
    df = pd.DataFrame({"id": [1]})
    add_processed_timestamp(df)
    assert "processed_at" not in df.columns
